### DISCUSSION TITLE
Using overlay - or set environment variables - from encrypted file

### DISCUSSION THREAD

**Discussion by gar1t** [nushell/nushell#10988] [General]
I'd like to set some API keys as environment variables. I store these in encrypted files.

In bash I'd run something like this:

    $ source <( gpg -d secrets.gpg )

Is there an approach I can take in Nushell for this?

**Comment by fdncred** [comment]
One place to start would be looking at `load-env` and `export-env`.

**Comment by gar1t** [comment]
Indeed, I'm using `load-env` in the non-encrypted module along with `overlay use`. This works fine but I can't keep secrets in plain text. What I'm missing is how to incorporate the keys from a gpg encrypted source.

**Comment by ghost** [comment]
### if you want compatibility with bash, docker, etc:

#### method1: a custom function

a function i use for `.env` files and similar (a bunch of lines in `key=value` format) is this:
```nu
def-env load_bash_env [] {
  load-env (
    $in | lines
    | where $it =~ "^[a-zA-Z_0-9]+=.*"  # remove comments and invalid lines
    | each {$in | parse "{key}={value}" | get 0}  # parse the values
    | transpose -i | headers | get 0  # format the output for load-env
  )
}
```
and use it with `open .env | load_bash_env`

#### method2: use toml/… and hope it is compatible

`foo=bar` is somewhat toml compatible (but it might not like some stings, etc)  
therefore it is sometimes possibles to just `load-env (open foo.toml)`  
(and it should probably be possible to get any `.env` file toml compatible)

### if compatibility is not a concern:  

#### method1: nu script as env file

just write a nu script and `source` it (this can be dangerous if other people could edit it (via git/…))

#### method2: nuon/…

write a `.nuon` env file
```
{
  foo: "bar"
}
```
and load it with `load-env (open foo.nuon)`

### encryption

all of these use streams and it is possible to just put some encryption method like `gpg` in between or replace the open: `gpg -d secrets.gpg | load_bash_env` or `load-env (gpg -d secrets.gpg | from nuon)`

**Comment by gar1t** [comment]
@Jan9103 that's very helpful. I'm going with the nuon based store. For the record

    load-env (gpg -d secrets.nuon.gpg | from nuon)

where `secrets.nuon.gpg` is a gpg-encrypted `.noun` env file, as suggested.
